### PR TITLE
[fix][authorization] Fix multiple roles authorization

### LINKED ISSUE #16574: MultiRolesTokenAuthorizationProvider only authorizes first role (at least for superuser)
First of all, thank you all for your work! This seems an amazing tool and thats why I am looking into it to use it as a domain event bus for my application.

**Describe the bug**
I'm setting JWT token authentication/authorization with multiple roles and using org.apache.pulsar.broker.authorization.MultiRolesTokenAuthorizationProvider in order to check for 'superuser' role in the list of roles that the JWT token has. It only works if the first role of the list is the 'superuser' one but do not work if there are other roles first.

I have not tested this with other roles applied on tenants or namespaces yet because of time but I may do It in the following days. Anyway, it seems is a common issue for any role but I'm not sure why this is happening viewing the code of AuthorizationProvider and related classes.

**To Reproduce**
I am building a more complex environment using Docker Swarm but for this kind of problem I have set a standalone configuration to reproduce it.

I have followed the steps on [https://pulsar.apache.org/ja/docs/standalone/](https://pulsar.apache.org/ja/docs/standalone/) with version `2.10.1` and modified the `standalone.conf` in order to set up JWT authentication/authorization following:

- https://pulsar.apache.org/docs/security-jwt
- https://pulsar.apache.org/docs/security-authorization#authorize-an-authenticated-client-with-multiple-roles

Following are the properties I have modified on `standalone.conf` file:
```
# Enable authentication
authenticationEnabled=true

# Authentication provider name list, which is comma separated list of class names
authenticationProviders=org.apache.pulsar.broker.authentication.AuthenticationProviderToken

# Enforce authorization
authorizationEnabled=true

# Authorization provider fully qualified class-name
authorizationProvider=org.apache.pulsar.broker.authorization.MultiRolesTokenAuthorizationProvider

superUserRoles=superuser

brokerClientAuthenticationPlugin=org.apache.pulsar.client.impl.auth.AuthenticationToken
brokerClientAuthenticationParameters={"token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJyb2xlcyI6WyJzdXBlcnVzZXIiLCJ0ZXN0Il19.fwFySHYsYES_j4ggOwShLJFsYiLBP9Ng0note_bex8Q"}

tokenSecretKey=data:;base64,dGVzdHNlY3JldFRvb29Mb29vb25nVG9CZVJlbWVtYmVyZWRGb3JUaGVTYWtlT2ZTaW1wbGljaXR5VW5leHBlY3RlZFdoYXRldmVyMQ==

tokenAuthClaim=roles
```

When I launch the standalone server using `bin/pulsar standalone` command and I try to access `http://localhost:8080/admin/v2/brokers/configuration` I get 401 error response as expected:
```sh
curl "http://localhost:8080/admin/v2/brokers/configuration"

<html>
<head>
<meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-1"/>
<title>Error 401 Authentication required</title>
</head>
<body><h2>HTTP ERROR 401 Authentication required</h2>
<table>
<tr><th>URI:</th><td>/admin/v2/brokers/configuration</td></tr>
<tr><th>STATUS:</th><td>401</td></tr>
<tr><th>MESSAGE:</th><td>Authentication required</td></tr>
<tr><th>SERVLET:</th><td>org.glassfish.jersey.servlet.ServletContainer-5ab2c5d8</td></tr>
</table>
<hr/><a href="https://eclipse.org/jetty">Powered by Jetty:// 9.4.44.v20210927</a><hr/>

</body>
</html>
```

If I build a JWT token with this claims:
```json
"roles": [
    "superuser",
    "test"
  ]
```

I get a successful response, as expected:
```sh
curl "http://localhost:8080/admin/v2/brokers/configuration" \
-H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJyb2xlcyI6WyJzdXBlcnVzZXIiLCJ0ZXN0Il19.fwFySHYsYES_j4ggOwShLJFsYiLBP9Ng0note_bex8Q"

["loadBalancerMemoryResourceWeight","ma.....
```

But, if I generate a JWT token with a change in the ordering:
```json
"roles": [
    "test",
    "superuser"
  ]
```

I get an unauthorized response:
```sh
curl "http://localhost:8080/admin/v2/brokers/configuration" \
-H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJyb2xlcyI6WyJ0ZXN0Iiwic3VwZXJ1c2VyIl19.Xz1gplVyg0MwTCobMlyKCu_xBLmMGCKYsTrvUyj9kL4"

{"reason":"This operation requires super-user access"}
```

**Expected behavior**
Setting `org.apache.pulsar.broker.authorization.MultiRolesTokenAuthorizationProvider` as the AuthorizationProvider with token authentication should authorize requests if the JWT token contains any of the `superUserRoles` in its `tokenAuthClaim` list.

**Desktop (please complete the following information):**
 - OS: macOS Monterey
 - Hardware: Mac mini (M1, 2020), Apple M1, 16GB
 On the Docker Swarm environment it is also reproducible and is on a x86_64 architecture with Ubuntu Server 18.04 LTS.

Thanks for your time!

### PR DESCRIPTION
**Pull request by nodece**
Fixes #16574 

### Motivation

The `MultiRolesTokenAuthorizationProvider` cannot handle the superuser and tenant admin correctly. It only checks if the role passed in is a superuser, which is incorrect. The correct way is to get a role list from authentication data and then check if the role list contains a superuser.

### Modifications

- Override the `isSuperAdmin`
- Override the `validateTenantAdminAccess`
- Add test for the `MultiRolesTokenAuthorizationProvider`

### Documentation

- [x] `doc-not-needed`

### CODE REVIEW

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/MultiRolesTokenAuthorizationProvider.java [hunk 1]
```diff
@@ -80,58 +86,112 @@ public void initialize(ServiceConfiguration conf, PulsarResources pulsarResource
...
+    @Override
+    public CompletableFuture<Boolean> isSuperUser(String role, AuthenticationDataSource authenticationData,
+                                                  ServiceConfiguration serviceConfiguration) {
+        Set<String> roles = getRoles(authenticationData);
+        if (roles.isEmpty()) {
+            return CompletableFuture.completedFuture(false);
+        }
+        Set<String> superUserRoles = serviceConfiguration.getSuperUserRoles();
+        if (superUserRoles.isEmpty()) {
+            CompletableFuture.completedFuture(false);
```
Comment (Technoboy-) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/MultiRolesTokenAuthorizationProvider.java:98]: Miss `return `

Comment (nodece) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/MultiRolesTokenAuthorizationProvider.java]: Good catch.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/MultiRolesTokenAuthorizationProviderTest.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+import org.apache.pulsar.client.admin.PulsarAdmin;
+import org.apache.pulsar.client.admin.PulsarAdminBuilder;
+import org.apache.pulsar.client.admin.PulsarAdminException;
+import org.apache.pulsar.client.impl.auth.AuthenticationToken;
+import org.apache.pulsar.common.policies.data.ClusterData;
+import org.apache.pulsar.common.policies.data.TenantInfo;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+public class MultiRolesTokenAuthorizationProviderTest extends MockedPulsarServiceBaseTest {
```
Comment (RobertIndie) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/api/MultiRolesTokenAuthorizationProviderTest.java:50]: Let's move this test to here: https://github.com/apache/pulsar/blob/master/pulsar-broker-common/src/test/java/org/apache/pulsar/broker/authorization/MultiRolesTokenAuthorizationProviderTest.java

**Review by RobertIndie — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/MultiRolesTokenAuthorizationProviderTest.java (hunk 2 above)
Comment (nodece) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/api/MultiRolesTokenAuthorizationProviderTest.java:50]: We cannot move this test to the pulsar-broker-common module, which is an integration test and depends on the pulsar-broker module.

Comment (RobertIndie) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/api/MultiRolesTokenAuthorizationProviderTest.java:50]: Got it. Thanks.

**Review by RobertIndie — APPROVED** [review]

**Review by coderzc — APPROVED** [review]

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/MultiRolesTokenAuthorizationProvider.java [hunk 3]
```diff
@@ -80,58 +86,112 @@ public void initialize(ServiceConfiguration conf, PulsarResources pulsarResource
...
+                                    throw new RestException(Response.Status.NOT_FOUND, "Tenant does not exist");
+                                }
+                            }).exceptionally(ex -> {
+                                Throwable cause = ex.getCause();
+                                if (cause instanceof MetadataStoreException.NotFoundException) {
+                                    log.warn("Failed to get tenant info data for non existing tenant {}", tenantName);
+                                    throw new RestException(Response.Status.NOT_FOUND, "Tenant does not exist");
+                                }
+                                log.error("Failed to get tenant {}", tenantName, cause);
+                                throw new RestException(cause);
```
Comment (Technoboy-) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/MultiRolesTokenAuthorizationProvider.java:138]: If line 129 throws RestException, will here wrap to  new RestException(RestException ...) ?

Comment (nodece) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/MultiRolesTokenAuthorizationProvider.java:138]: RestException can get the real exception.

```java
public RestException(Throwable t) {
        super(getResponse(t));
    }

    private static Response getResponse(Throwable t) {
        if (t instanceof WebApplicationException) {
            WebApplicationException e = (WebApplicationException) t;
            return e.getResponse();
        } else {
            return Response
                .status(Status.INTERNAL_SERVER_ERROR)
                .entity(getExceptionData(t))
                .type(MediaType.TEXT_PLAIN)
                .build();
        }
    }
```

**Comment by BewareMyPower** [comment]
This PR relies on the `TenantResources#getTenantAsync` method, which relies on the `BaseResources#joinPath` method that I tried to migrate in https://github.com/apache/pulsar/pull/16867. Could you help review that PR so that I can continue cherry-picking and resolving the conflicts?

**Comment by nodece** [comment]
@BewareMyPower Done.

**Pull request merged**